src/H5Tconv.c

```diff
@@ -2753,131 +2795,205 @@ H5T__conv_struct_opt(H5T_t *src, H5T_t *dst, H5T_cdata_t *cdata, const H5T_conv_
  *-------------------------------------------------------------------------
  */
 static herr_t
-H5T__conv_enum_init(H5T_t *src, H5T_t *dst, H5T_cdata_t *cdata)
-{
-    H5T_enum_struct_t *priv = NULL;         /*private conversion data    */
-    int                n;                   /*src value cast as native int    */
-    int                domain[2] = {0, 0};  /*min and max source values    */
-    int               *map       = NULL;    /*map from src value to dst idx    */
-    unsigned           length;              /*nelmts in map array        */
-    unsigned           i, j;                /*counters            */
-    herr_t             ret_value = SUCCEED; /* Return value */
+H5T__conv_enum_init(const H5T_t *src, const H5T_t *dst, H5T_cdata_t *cdata, const H5T_conv_ctx_t *conv_ctx)
+{
+    H5T_conv_enum_t *priv          = NULL; /* Private conversion data */
+    int             *map           = NULL; /* Map from src value to dst idx */
+    bool             rebuild_cache = false;
+    herr_t           ret_value     = SUCCEED;
 
     FUNC_ENTER_PACKAGE
 
     cdata->need_bkg = H5T_BKG_NO;
-    if (NULL == (priv = (H5T_enum_struct_t *)(cdata->priv = H5MM_calloc(sizeof(*priv)))))
-        HGOTO_ERROR(H5E_RESOURCE, H5E_NOSPACE, FAIL, "memory allocation failed");
-    if (0 == src->shared->u.enumer.nmembs)
-        HGOTO_DONE(SUCCEED);
 
-    /*
-     * Check that the source symbol names are a subset of the destination
-     * symbol names and build a map from source member index to destination
-     * member index.
-     */
-    H5T__sort_name(src, NULL);
-    H5T__sort_name(dst, NULL);
-    if (NULL == (priv->src2dst = (int *)H5MM_malloc(src->shared->u.enumer.nmembs * sizeof(int))))
-        HGOTO_ERROR(H5E_RESOURCE, H5E_NOSPACE, FAIL, "memory allocation failed");
-    for (i = 0, j = 0; i < src->shared->u.enumer.nmembs && j < dst->shared->u.enumer.nmembs; i++, j++) {
-        while (j < dst->shared->u.enumer.nmembs &&
-               strcmp(src->shared->u.enumer.name[i], dst->shared->u.enumer.name[j]) != 0)
-            j++;
-        if (j >= dst->shared->u.enumer.nmembs)
-            HGOTO_ERROR(H5E_DATATYPE, H5E_UNSUPPORTED, FAIL,
-                        "source type is not a subset of destination type");
-        priv->src2dst[i] = (int)j;
-    } /* end for */
+    priv = (H5T_conv_enum_t *)(cdata->priv);
+    if (!priv) {
+        if (NULL == (priv = (H5T_conv_enum_t *)(cdata->priv = calloc(1, sizeof(*priv)))))
+            HGOTO_ERROR(H5E_RESOURCE, H5E_NOSPACE, FAIL, "memory allocation failed");
+        if (0 == src->shared->u.enumer.nmembs)
```

Comment:
This 'if' (on the source # of members) can be removed, since it's covered in the 'if(rebuil_cache)' case.

Reply:
True. Will remove.